backend/app/analysis/yolo_pop_density.py

```diff
@@ -0,0 +1,108 @@
+    else:
+        line_overlap += 1
+    return line_overlap / 2
+
+
+def overlap_2d(rect1, rect2):
+    """
+    Given a two rectangles, return a float representing how overlapped they are.
+    :param rect1: A list/tuple of (x_position, y_position, width, height)
+    :param rect2: A list/tuple of (x_position, y_position, width, height)
+    :return: A float from 1 - 0 representing how overlapped the two input rectangles are
+    (A value of ~.5 indicates that the lines are approximately touching but not overlapped.)
+    """
+    x1, y1, w1, h1 = rect1
+    x2, y2, w2, h2 = rect2
+    x_overlap = overlap_1d((x1, x1 + w1), (x2, x2 + w2))
+    y_overlap = overlap_1d((y1, y1 + h1), (y2, y2 + h2))
+    return x_overlap * y_overlap
+
+
+def box_to_rect(box):
+    """
+    Given a box representing an object's location in a photo as predicted by the yolo_model,
+    return a tuple representing a rectangle to be used for calculating overlap.
+    :param box: dictionary with keys 'x_coord', 'y_coord', 'width', 'height'
+    :return: tuple (rectangle)
+    """
+    return (
+        box['x_coord'],
+        box['y_coord'],
+        box['width'],
+        box['height']
+    )
+
+
+def object_density(object_name, yolo_dict, photo_dim=(10000, 10000)):
+    """
+    Given an object name, the output of the yolo_model analysis for a photo (yolo_dict),
+    and the dimensions of the photo, return the density of the specified object
+    in the photo.
+    :param object_name: string (name of the object)
+    :param yolo_dict: dictionary (output of the yolo_model for the photo)
+    :param photo_dim: tuple (dimensions of the photo the object is located in)
+    :return: A number representing how densely the specified object type is packed
+    in an image. More overlapping objects corresponds to higher density.
+    """
+    boxes = *filter(lambda box: box['label'] == object_name, yolo_dict.get("boxes", {})),
+    if len(boxes) == 1:
+        photo_width, photo_height = photo_dim
+        rect1 = box_to_rect(boxes[0])
+        rect2 = (
+            boxes[0]['x_coord'] + photo_width,  # Assume that there is another object
+            boxes[0]['y_coord'] + photo_height,  # outside the frame of the photo
+            boxes[0]['width'],
+            boxes[0]['height']
+        )
+        return overlap_2d(rect1, rect2) + overlap_2d(rect2, rect1)
+    density = 0
+    for i in range(len(boxes)):
+        density_i = 0
+        for j in range(len(boxes)):
+            if i == j:
+                continue
+            density_i += overlap_2d(
+                box_to_rect(boxes[i]), box_to_rect(boxes[j])
+            )
+        density += density_i / (len(boxes) - 1)
+    return density
+
+
+def analyze(photo: Photo):
+    yolo_dict = PhotoAnalysisResult.objects.filter(
+        name="yolo_model", photo=photo
+    ).first().parsed_result()
+    image = photo.get_image_data()
+    photo_dim = image.shape[:2][::-1] if image else (10000, 10000)
```

Comment:
`if image` fails here under some conditions with this error from numpy:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Basically: the numpy notion here is that whether an array is truthy or not is ambiguous, as a user of the library might reasonably assume that an array is truthy if _all_ of its elements are truthy or if any elements exist (as in a Python list). What we probably want here is actually to if on the size, or `if image is not None`.